### PR TITLE
EXSWHTEC-284 - Implement tests for square/cube root device math functions

### PR DESCRIPTION
**Pull request by nives-vukovic**
Depends-on: #260 
- Accuracy tests for single-precision and double-precision sqrt/rsqrt/cbrt/rcbrt/hypot/rhypot/norm3d/rnorm3d/norm4d/rnorm4d
- Sanity tests for single-precision and double-precision norm/rnorm
- Negative tests for single-precision and double precision square/cube root math functions

### CODE REVIEW

File: catch/unit/math/math_common.hh [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    kernel_<<<grid_dim, block_dim>>>(y_dev_.ptr(), num_args, std::get<I>(xss_dev_).ptr()...);
+    HIP_CHECK(hipGetLastError());
+
+    HIP_CHECK(hipMemcpy(y_.ptr(), y_dev_.ptr(), num_args * sizeof(T), hipMemcpyDeviceToHost));
+    HIP_CHECK(hipStreamSynchronize(nullptr));
+
+    if constexpr (!parallel) {
+      for (auto i = 0u; i < num_args; ++i) {
+        const auto actual_val = y_.ptr()[i];
+        const auto ref_val = static_cast<T>(ref_func(xss[i]...));
```
Comment (b-sumner) [inline catch/unit/math/math_common.hh:123]: Just an observation that this cast and other similar ones are eliminating the ability to check for, or report fractional ulp error values.

File: catch/unit/math/root_funcs.cc [hunk 2]
```diff
@@ -0,0 +1,600 @@
...
+  LinearAllocGuard<T> x{LinearAllocs::hipHostMalloc, max_dim * sizeof(T)};
+  LinearAllocGuard<T> x_dev{LinearAllocs::hipMalloc, max_dim * sizeof(T)};
+  LinearAllocGuard<T> y{LinearAllocs::hipHostMalloc, sizeof(T)};
+  LinearAllocGuard<T> y_dev{LinearAllocs::hipMalloc, sizeof(T)};
+
+  std::fill_n(x.ptr(), max_dim, 1);
+  HIP_CHECK(hipMemcpy(x_dev.ptr(), x.ptr(), max_dim * sizeof(T), hipMemcpyHostToDevice));
+
+  for (uint64_t i = 1u; i < max_dim; i++) {
+    kernel<<<grid_size, block_size>>>(y_dev.ptr(), i, x_dev.ptr());
```
Comment (b-sumner) [inline catch/unit/math/root_funcs.cc]: Why do we need more than 1 thread if they all compute the same result?

**Review by b-sumner — COMMENTED** [review]
This is way to much to review at once.  I get that the infrastructure would be large, but everything else could have been split into smaller chunks.

File: catch/unit/math/math_common.hh (hunk 1 above)
Comment (nives-vukovic) [reply catch/unit/math/math_common.hh:123]: @b-sumner There is a separate PR for common numerical accuracy files: https://github.com/ROCm-Developer-Tools/hip-tests/pull/233 It is best to review and comment these code segments on this PR, as all the others depend on it.

File: catch/unit/math/root_funcs.cc (hunk 2 above)
Comment (nives-vukovic) [reply catch/unit/math/root_funcs.cc]: @b-sumner Yes, it is supposed to be only one thread, fixed.

**Comment by b-sumner** [comment]
Related to my first comment, I would like to request that these reviews be reorganized into smaller pieces starting with core test infrastructure changes and followed by reviews for each new added test or highly related family of test.  Ideally the core changes would have been approved before new test development, but now any core changes needed will have to be propagated to all the tests. Ideally an individual review would touch no more than 3-4 files and fit on no more than 2-3 screens.

**Comment by nives-vukovic** [comment]
@b-sumner Thank you for your review. We tried to do exactly that by creating #233 , all the PRs (including this one) contain these common core files, but as the core PR is not yet merged it is hard to go through each of them to find specific files that are not core files. In this PR those are root_funcs.cc with main implementation for root and cubic math functions and math_root_negative* which are used to simply check if compilation fails when functions arguments are of invalid type (that same approach has been used in atomic PRs).

**Comment by b-sumner** [comment]
#233 has 17 files which by itself is a concern.  And including them in every subsequent review is a bigger concern.  This review process would have gone more smoothly if you could restrict to a handful of files and wait for merge to submit the next handful.  What we have now with several reviews each with 15+ files, some of which may or may not be identical is a problem so I would like to back up and try again with the smooth approach.

**Comment by mangupta** [comment]
@nives-vukovic : Could you please look into splitting this PR into smaller ones so that it can be reviewed easily?

**Comment by nives-vukovic** [comment]
@mangupta This PR depends on PR #260, which is also chained with another PRs. When all PRs in the dependency chain are approved and merged, this PR will not be very big and will be simple to review. All math PRs have a common base which is large, therefore this was the only viable approach.

**Comment by rakesroy** [comment]
PR has been merged into develop branch via commit 46ada257307db5e604948478b6325092121e29c1.

**Pull request closed without merging**